### PR TITLE
Support structs for `cudf::contains` with column/scalar input

### LINKED ISSUE #8965: [FEA] Add nested struct support for cudf::contains
**Is your feature request related to a problem? Please describe.**
For Spark we are pushing to get more support for structs in a number of operators. We already have some support for sorting structs, so we should be able to come up with a way to check for existence of a struct value from one column to another. NOTE this does not include lists as children of the structs just structs that contains basic types including strings and other structs.

his should follow the same pattern we have supported for sorting where null child columns are considered equal to other null child columns.  Like described in https://github.com/rapidsai/cudf/issues/8964

**Describe the solution you'd like**
I would like to see cudf::contains updated so it can support this.

**Describe alternatives you've considered**
I don't think there is an alternative that we can do on our own.

### PR DESCRIPTION
**Pull request by ttnghia**
This PR adds support for `cudf::contains` so we can check whether a structs column contains a scalar struct element.

Partially addresses #8965. This does not support checking if structs given in a structs column exist in another structs column. Such cases will be supported when the new data structure mentioned in https://github.com/rapidsai/cudf/issues/9413 is merged into cudf.

### CODE REVIEW

**Comment by hyperbolic2346** [comment]
Copyrights

**Review by hyperbolic2346 — APPROVED** [review]

File: cpp/src/search/search.cu [hunk 1]
```diff
@@ -173,11 +173,50 @@ bool contains_scalar_dispatch::operator()<cudf::list_view>(column_view const&,
...
-                                                             scalar const&,
-                                                             rmm::cuda_stream_view)
+bool contains_scalar_dispatch::operator()<cudf::struct_view>(column_view const& col,
+                                                             scalar const& value,
+                                                             rmm::cuda_stream_view stream)
 {
-  CUDF_FAIL("struct_view type not supported yet");
+  CUDF_EXPECTS(col.type() == value.type(), "scalar and column types must match");
+  auto const scalar_table = static_cast<struct_scalar const*>(&value)->view();
+  CUDF_EXPECTS(col.num_children() == scalar_table.num_columns(),
```
Comment (mythrocks) [inline cpp/src/search/search.cu:182]: Should we not also check whether the column's children match types with the scalar's children?

File: cpp/src/search/search.cu [hunk 2]
```diff
@@ -173,11 +173,50 @@ bool contains_scalar_dispatch::operator()<cudf::list_view>(column_view const&,
...
+  auto const val_flattened =
+    structs::detail::flatten_nested_columns(scalar_table, {}, {}, flatten_nullability);
+
+  // The struct scalar only contains the struct member columns.
+  // Thus, if there is any null in the input, we must exclude the first column in the flattenned
+  // table of the input column from searching because that column is the materialized bitmask of
+  // the input structs column.
+  auto const col_flattened_content  = col_flattened.flattened_columns();
+  auto const col_flattened_children = table_view{std::vector<column_view>{
+    col_flattened_content.begin() + has_null_elements, col_flattened_content.end()}};
```
Comment (mythrocks) [inline cpp/src/search/search.cu:205]: :+1: This is clever. I like this phrasing, although I remember a warning about using boolean values as integers.

File: cpp/src/search/search.cu [hunk 3]
```diff
@@ -173,11 +173,50 @@ bool contains_scalar_dispatch::operator()<cudf::list_view>(column_view const&,
...
+                                     : structs::detail::column_nullability::MATCH_INCOMING;
+
+  // Flatten the input structs column, only materialize the bitmask if there is null in the input.
+  auto const col_flattened =
+    structs::detail::flatten_nested_columns(table_view{{col}}, {}, {}, flatten_nullability);
+  auto const val_flattened =
+    structs::detail::flatten_nested_columns(scalar_table, {}, {}, flatten_nullability);
+
+  // The struct scalar only contains the struct member columns.
+  // Thus, if there is any null in the input, we must exclude the first column in the flattenned
```
Comment (mythrocks) [inline cpp/src/search/search.cu:200]:
```suggestion
  // Thus, if there is any null in the input, we must exclude the first column in the flattened
```

**Review by mythrocks — APPROVED** [review]
LGTM, apart from a minor nitpick.

File: cpp/src/search/search.cu (hunk 2 above)
Comment (ttnghia) [reply cpp/src/search/search.cu:205]: Should not be any warning here, but I'll change that for clarity anyway. Thanks.

**Comment by ttnghia** [comment]
@gpucibot merge

**Pull request merged**